Add ability to change language without a restart

rubenwardy opened the pull request:
This allows changing the current language without restarting Minetest. Gettext reads the `language` env var for every call, so we just need to change this and notify gettext about the change

## To do

This PR is  Ready for Review.

## How to test

* Set system language to non-English, ie French
* Clear minetest.conf
* Open Minetest, see in French language
* Change Minetest language setting to German. See it change to German
* Change Minetest language setting back to (Use system language). See French

y5nw commented:
Two minor issues I found:

1. Switching from a specific language to "[Use system language]" does not seem to apply the language.
2. Changing the language when MT is run using `LC_ALL=... bin/minetest` messes up the UI (see attachment)

![2024-02-17-18-56-02](https://github.com/minetest/minetest/assets/37980625/3556875a-988e-4e97-92ba-46afd2b52840)

sfan5 commented on src/gettext.cpp:
```diff
@@ -216,30 +249,12 @@ void init_gettext(const char *path, const std::string &configured_language,
+	// Source: https://www.gnu.org/software/gettext/manual/html_node/gettext-grok.html
+	{
+		extern int  _nl_msg_cat_cntr;
+		++_nl_msg_cat_cntr;
```
I wonder if this won't mess up on systems where they are using some sort of alternative gettext implementation

rubenwardy replied:
Whilst the documentation says to do this, I found that it wasn't actually needed.  So if it's likely to cause issues, could maybe be removed

y5nw replied:
Related: https://bugs.kde.org/show_bug.cgi?id=365917 (fixed by https://invent.kde.org/frameworks/ki18n/-/commit/543b8ae67c57500bf3cabde75bf1599904d6e482) and also the https://github.com/minetest/minetest/pull/14379#issuecomment-1975327822 comment below.

Maybe only enable language switching when it can be determined that it works? (e.g. by checking whether translating LANG_CODE works after switching languages)

rubenwardy commented on src/gettext.cpp:
```diff
@@ -216,30 +249,12 @@ void init_gettext(const char *path, const std::string &configured_language,
-	/* no matter what locale is used we need number format to be "C" */
-	/* to ensure formspec parameters are evaluated correctly!        */
-
-	setlocale(LC_NUMERIC, "C");
```
Thanks for testing, thought this might be the case

grorp commented:
What do you think about using a `SettingsChangedCallback` instead of manually reporting changes via `core.set_language`? This way, I imagine that changing the setting via a chat command in singleplayer (e.g. `/set language it`) could also work.

rubenwardy commented:
> What do you think about using a `SettingsChangedCallback` instead of manually reporting changes via `core.set_language`? This way, I imagine that changing the setting via a chat command in singleplayer (e.g. `/set language it`) could also work.

Yeah I was thinking about this, but decided to go with the simpler option. But I'm not set on it. It would make the settings component clearer, no need to add a `language` specific implementation to it

One thing to point out is that `/set` does the _server-side_ setting, so may only work in singleplayer (assuming set_gettext_language is thread safe). I don't know if there's a client-side equivalent to `/set` (`.set` ?). In any case, having a language setting in the pause menu (#6722) would be very welcome

rubenwardy commented:
Updated, please may you retest @y5nw

y5nw commented on src/gettext.cpp:
```diff
@@ -164,12 +164,52 @@ static void MSVC_LocaleWorkaround(int argc, char* argv[])
 #if USE_GETTEXT
-	// First, try to set user override environment
+	char *system_language_c = getenv("LANGUAGE");
+	if (system_language_c == NULL)
```
This should also check for the empty string and the value of `LC_ALL`.

rubenwardy replied:
I can't get LC_ALL to work in the master branch, so it appears that it's ignored when it comes to detecting the system language

```
unset LANGUAGE
unset LANG
LC_ALL=de_DE.UTF-8 ./minetest  --verbose
```

y5nw replied:
I seem to get very interesting results with this.

My current setup has `en_US` and the `de_DE` locales (and some more), but not `fr_FR` listed in `locale -a`, and:

```sh
# LANGUAGE= (empty string), LANG=en_US.UTF-8
env LC_ALL=de_DE.UTF-8 bin/minetest # UI appears in German, same for LANG=... and LANGUAGE=...
env LC_ALL=fr_FR.UTF-8 bin/minetest # UI appears in English
env LANG=fr_FR.UTF-8 bin/minetest # UI appears in English
env LANGUAGE=fr_FR.UTF-8 bin/minetest # UI appears in French

unset LANGUAGE
# same as above

unset LANG
env LC_ALL=de_DE.UTF-8 bin/minetest # UI appears in German, same for LANG=...
env LANGUAGE=de_DE.UTF-8 bin/minetest # UI appears in English (!)
env LC_ALL=fr_FR.UTF-8 bin/minetest # UI appears in English
env LANG=fr_FR.UTF-8 bin/minetest # UI appears in English
env LANGUAGE=fr_FR.UTF-8 bin/minetest # UI appears in English

export LANGUAGE=
# same as above
```

So (at least on my computer) it seems like setting `LC_ALL` or `LANG` to a language not listed in `locale -a` does not work, while setting `LANGUAGE` only works when `LANG` is set to a known (?) locale.

It should also be noted that `env LC_ALL=fr_FR.UTF-8 locale` (and `env LANG=fr_FR.UTF-8 locale`) throws the errors
```
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
```
while `env LANGUAGE=fr_FR.UTF-8 locale` does not.

More interestingly, the result is the same when using `gettext --version` instead of `bin/minetest`, so 🫠

That said, I would assume this is somewhat related to the first entry in https://www.gnu.org/software/gettext/FAQ.html#integrating_noop.

IMO the sane behavior here would be to set the language based on `LANGUAGE`, `LC_ALL`,  `LANG` (in that order) and let gettext figure out the rest.

---

Some more information on my setup:
-  Arch Linux
- `gettext --version`: `gettext (GNU gettext-runtime) 0.22.4`
- `locale --version`: `locale (GNU libc) 2.38`

rubenwardy replied:
> So (at least on my computer) it seems like setting LC_ALL or LANG to a language not listed in locale -a does not work, while setting LANGUAGE only works when LANG is set to a known (?) locale.

Ahhh that'll be why, my `locale -a` only has C, en, and POSIX. I'll see how to install a language so I can test this. Thanks for looking into this

rubenwardy replied:
Given that we only ever set LANGUAGE, it might be enough to just reset that env var and rely on gettext's logic on reading other env vars. I'll work out how to test LC_ALL and see if this approach works

sfan5 replied:
we might have to unset the other variables however depending on priority order

y5nw replied:
It seems like (re)setting `LANGUAGE` alone is enough, and gettext does read other environment variables when `LANGUAGE` is set to the empty string. (At least it works for me)

That said, I do not have a Windows machine to test the MSVC workaround, so I don't know what happens when `LANGUAGE` is reset to the empty string there.

> we might have to unset the other variables however depending on priority order

`LANGUAGE` is [at the top of the priority order](https://www.gnu.org/software/gettext/manual/html_node/Locale-Environment-Variables.html) so that is not a problem.

rubenwardy commented:
Ok, updated. It works as expected for me with `LANGUAGE`, although I'm unable to test `LC_ALL`

y5nw commented:
> Ok, updated. It works as expected for me with `LANGUAGE`, although I'm unable to test `LC_ALL`

Using `LC_ALL` also works.

rubenwardy commented:
Rebased. As the ContentDB dialog now requests localised meta, I've added a `reset_contentdb()` function to clear the packages when changing the language

rubenwardy commented:
@wsor4035 tested on Windows using the mingw and MSVC builds from GitHub actions. The Mingw builds pass the test plan in the opening post, including detecting system language. The MSVC build is not compiled with gettext so does not support translation

MSVC left, mingw right

![image](https://github.com/minetest/minetest/assets/2122943/3ee97815-49d6-4bcf-93cb-efc7bfe438bf)

y5nw commented:
MSVC appears to be particularly hard to work with (at least from what I tried in a VM).

It seems like `gettext[tools]` is needed to build Minetest with MSVC and Gettext. I was able to _build_ this with the following patch:

```diff
diff --git a/src/gettext.cpp b/src/gettext.cpp
index 972519116..8825d518d 100644
--- a/src/gettext.cpp
+++ b/src/gettext.cpp
@@ -34,6 +34,9 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define setlocale(category, localename) \
 	setlocale(category, MSVC_LocaleLookup(localename))
 
+static int argc;
+static char** argv;
+
 static std::map<std::wstring, std::wstring> glb_supported_locales;
 
 /******************************************************************************/
@@ -115,7 +118,7 @@ static const char* MSVC_LocaleLookup(const char* raw_shortname)
 	return "";
 }
 
-static void MSVC_LocaleWorkaround(int argc, char* argv[])
+static void MSVC_LocaleWorkaround()
 {
 	errorstream << "MSVC localization workaround active.  "
 		"Restarting " PROJECT_NAME_C " in a new environment!" << std::endl;
@@ -183,7 +186,7 @@ namespace {
 
 /******************************************************************************/
 void init_gettext(const char *path, const std::string &configured_language,
-	int argc, char *argv[])
+	int _argc, char *_argv[])
 {
 #if USE_GETTEXT
 	char *system_language_c = getenv("LANGUAGE");
@@ -205,6 +208,10 @@ void init_gettext(const char *path, const std::string &configured_language,
 	assert(tdomain);
 	if (tdomain)
 		bind_textdomain_codeset(tdomain, "UTF-8");
+#ifdef _MSC_VER
+	argc = _argc;
+	argv = _argv;
+#endif
 #endif
 
 #else
@@ -254,21 +261,23 @@ void set_gettext_language(std::string configured_language)
 #ifndef SERVER
 	// Hack to force gettext to see the right environment
 	if (current_language != configured_language)
-		MSVC_LocaleWorkaround(argc, argv);
+		MSVC_LocaleWorkaround();
 #else
 	errorstream << "*******************************************************" << std::endl;
 	errorstream << "Can't apply locale workaround for server!" << std::endl;
 	errorstream << "Expect language to be broken!" << std::endl;
 	errorstream << "*******************************************************" << std::endl;
 #endif
-#endif
+#else
 
 	// Notify gettext that the language has changed
 	// Source: https://www.gnu.org/software/gettext/manual/html_node/gettext-grok.html
+	// This does not seem to work when building with MSVC
 	{
 		extern int  _nl_msg_cat_cntr;
 		++_nl_msg_cat_cntr;
 	}
+#endif
 #endif
 
 	/* no matter what locale is used we need number format to be "C" */
```

However, switching the language causes MT to crash.

![2024-03-03-22-03-58](https://github.com/minetest/minetest/assets/37980625/5d310fe0-8c37-4f69-bf8e-2606b419265d)

This is the stack trace as reported by Visual Studio 2022:
```
 	ucrtbased.dll!00007fffc81b2fb5()	未知
 	ucrtbased.dll!00007fffc81b3153()	未知
 	ucrtbased.dll!00007fffc81cae2d()	未知
 	msvcp140d.dll!00007fffcbd426b1()	未知
 	msvcp140d.dll!00007fffcbd423e2()	未知
>	minetest.exe!std::_Mutex_base::~_Mutex_base() 行 50	C++
 	minetest.exe!std::mutex::~mutex()	C++
 	minetest.exe!Settings::~Settings() 行 131	C++
 	[外部代码]	
 	minetest.exe!uninit_common() 行 726	C++
 	[外部代码]	
 	minetest.exe!MSVC_LocaleWorkaround() 行 150	C++
 	minetest.exe!set_gettext_language(std::string configured_language) 行 285	C++
 	minetest.exe!<lambda_69150f7968cdfca1ce97613eb7c6d8d4>::operator()<std::string,void *>(std::string _name, void * _data) 行 712	C++
 	minetest.exe!<lambda_69150f7968cdfca1ce97613eb7c6d8d4>::<lambda_invoker_cdecl><std::string const &,void *>(const std::string & _name, void * _data) 行 713	C++
 	minetest.exe!Settings::doCallbacks(const std::string & name) 行 1092	C++
 	minetest.exe!Settings::set(const std::string & name, const std::string & value) 行 871	C++
 	minetest.exe!LuaSettings::l_set(lua_State * L) 行 205	C++
 	minetest.exe!script_exception_wrapper(lua_State * L, int(*)(lua_State *) f) 行 41	C++
 	[外部代码]	
 	lua51.dll!lua_pcall(lua_State * L, int nargs, int nresults, int errfunc) 行 1145	C
 	minetest.exe!ScriptApiMainMenu::handleMainMenuButtons(const std::unordered_map<std::string,std::string,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::pair<std::string const ,std::string>>> & fields) 行 91	C++
 	minetest.exe!TextDestGuiEngine::gotText(const std::unordered_map<std::string,std::string,std::hash<std::string>,std::equal_to<std::string>,std::allocator<std::pair<std::string const ,std::string>>> & fields) 行 53	C++
 	minetest.exe!GUIFormSpecMenu::acceptInput(FormspecQuitMode quitmode) 行 4068	C++
 	minetest.exe!GUIFormSpecMenu::OnEvent(const irr::SEvent & event) 行 4999	C++
 	minetest.exe!irr::gui::IGUIElement::OnEvent(const irr::SEvent & event) 行 549	C++
 	minetest.exe!irr::gui::IGUIElement::OnEvent(const irr::SEvent & event) 行 549	C++
 	minetest.exe!GUIScrollContainer::OnEvent(const irr::SEvent & event) 行 57	C++
 	minetest.exe!irr::gui::CGUIComboBox::sendSelectionChangedEvent() 行 377	C++
 	minetest.exe!irr::gui::CGUIComboBox::OnEvent(const irr::SEvent & event) 行 298	C++
 	minetest.exe!irr::gui::CGUIListBox::selectNew(int ypos, bool onlyHover) 行 475	C++
 	minetest.exe!irr::gui::CGUIListBox::OnEvent(const irr::SEvent & event) 行 426	C++
 	minetest.exe!irr::gui::CGUIEnvironment::postEventFromUser(const irr::SEvent & event) 行 597	C++
 	minetest.exe!irr::CIrrDeviceStub::postEventFromUser(const irr::SEvent & event) 行 224	C++
 	minetest.exe!irr::CIrrDeviceSDL::run() 行 695	C++
 	minetest.exe!RenderingEngine::run() 行 149	C++
 	minetest.exe!GUIEngine::run() 行 329	C++
 	minetest.exe!GUIEngine::GUIEngine(JoystickController * joystick, irr::gui::IGUIElement * parent, RenderingEngine * rendering_engine, IMenuManager * menumgr, MainMenuData * data, bool & kill) 行 205	C++
 	minetest.exe!ClientLauncher::main_menu(MainMenuData * menudata) 行 551	C++
 	minetest.exe!ClientLauncher::launch_game(std::string & error_message, bool reconnect_requested, GameStartData & start_data, const Settings & cmd_args) 行 427	C++
 	minetest.exe!ClientLauncher::run(GameStartData & start_data, const Settings & cmd_args) 行 191	C++
 	minetest.exe!main(int argc, char * * argv) 行 265	C++
 	[外部代码]	
```

rubenwardy commented:
Does it work on master?

rubenwardy commented:
Yeah, I didn't read `MSVC_LocaleWorkaround` - restarting the app like that is incompatible with the changing at runtime thing. We should probably see if it's still needed

y5nw commented:
> Does it work on master?

MSVC+Gettext seems to work on the master branch, but without language switching.

> We should probably see if it's still needed - this workaround was added 11 years ago ( [22a59b3](https://github.com/minetest/minetest/commit/22a59b3912ff5e7bb1516faa06f1841545a8117c) ) and so might not be needed any more. I suggest commenting out MSVC_LocaleWorkaround and seeing if it works

I tried removing the workaround. The result is that changing the language only updates the translations for strings where a translation is previously not available:

![2024-03-03-22-17-20](https://github.com/minetest/minetest/assets/37980625/eb2805c6-5b8a-4677-8ba5-2777b81ff81d)
![2024-03-03-22-17-25](https://github.com/minetest/minetest/assets/37980625/6c546207-420f-4260-804a-c100eaa418c6)

That said, I removed the `_nl_msg_cat_cntr++` code when compiling with MSVC as it resulted in an error, so I am not sure how relevant that is.

IMO it would be best here to change `set_gettext_language` to no-op for MSVC for now unless someone comes up with a proper solution that does not involve something like this workaround.

y5nw commented on src/gettext.cpp:
```diff
@@ -232,14 +208,73 @@ void init_gettext(const char *path, const std::string &configured_language,
 #endif
```
It should be noted that the `_WIN32` check here should be removed since some locales on Linux also do not use a UTF-8 locale by default.
